internal/pkg/aws/cloudformation/cloudformation.go

```diff
@@ -0,0 +1,132 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package cloudformation provides a client to make API requests to AWS CloudFormation.
+package cloudformation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+)
+
+var waiters = []request.WaiterOption{
+	request.WithWaiterDelay(request.ConstantWaiterDelay(3 * time.Second)), // Poll for cfn updates every 3 seconds.
+	request.WithWaiterMaxAttempts(1800),                                   // Wait for at most 90 mins for any cfn action.
+}
+
+type api interface {
+	changeSetAPI
+
+	DescribeStacks(*cloudformation.DescribeStacksInput) (*cloudformation.DescribeStacksOutput, error)
+	DeleteStack(*cloudformation.DeleteStackInput) (*cloudformation.DeleteStackOutput, error)
+	WaitUntilStackDeleteCompleteWithContext(aws.Context, *cloudformation.DescribeStacksInput, ...request.WaiterOption) error
+}
+
+// CloudFormation represents a client to make requests to AWS CloudFormation.
+type CloudFormation struct {
+	client api
+}
+
+// New creates a new CloudFormation client.
+func New(s *session.Session) *CloudFormation {
+	return &CloudFormation{
+		client: cloudformation.New(s),
+	}
+}
+
+// Create deploys a new CloudFormation stack.
+func (c *CloudFormation) Create(stack *Stack) error {
+	descr, err := c.describe(stack.name)
+	if err != nil {
+		var stackNotFound *errStackNotFound
+		if !errors.As(err, &stackNotFound) {
+			return err
+		}
+		// If the stack does not exist, create it.
+		return c.createChangeSet(stack)
+	}
+	status := stackStatus(aws.StringValue(descr.StackStatus))
+	if status.requiresCleanup() {
+		// If the stack exists, but failed to create, we'll clean it up and then re-create it.
+		if err := c.delete(stack.name); err != nil {
+			return fmt.Errorf("cleanup previously failed stack %s: %w", stack.name, err)
+		}
+		return c.createChangeSet(stack)
+	}
+	if status.inProgress() {
+		return &errStackUpdateInProgress{
+			name: stack.name,
+		}
+	}
+	return &ErrStackAlreadyExists{
```

Comment:
Maybe in the future we can change this method from create to just CreateOrUpdate or Deploy or something, and just update if it already exists. Just thinking out loud, no need to change it now.

Reply:
I'm also thinking, for the env init idempotency thing, it might be nice to add the actual stack as a parameter to the error. So we can pull values out. 

Right now, CreateEnv expects an env to be returned - which is just constructed from a few values from the actual stack - so it'd be convenient for it to just catch this err, and create the env from the existing stack value.

Reply:
Gotcha! good idea, done!